corehq/apps/locations/fixtures.py

```diff
@@ -66,13 +66,23 @@ def should_sync_locations(last_sync, locations_queryset, restore_user):
     ):
         return True
 
-    return (
-        locations_queryset.filter(last_modified__gte=last_sync.date).exists()
-        or LocationType.objects.filter(domain=restore_user.domain,
-                                       last_modified__gte=last_sync.date).exists()
-    )
+    if (locations_queryset.filter(last_modified__gte=last_sync.date).exists()
+        or LocationType.objects.filter(
+            domain=restore_user.domain, last_modified__gte=last_sync.date).exists()):
+        return True
+
+    if toggles.RELATED_LOCATIONS.enabled(restore_user.domain):
+        return _location_relation_changed(locations_queryset, last_sync.date)
```

Comment:
Agreed that this is more expensive than expected. This does not change the number of dbs hit for non-reach pilot domains (toggle would be cached to the request), so its impact should be minimal for all other projects.

What I'm not sure about, is if we want to prioritize our caching to optimize database load, or caching to optimize network packet size since we try to optimize for low resource settings. I think there are projects with large location restores (ICDS does not at the moment, but I believe ANMs in the pilot may break that mold) so this would still be quite helpful for them

Reply:
Agreed, and I think it may be reasonable to do this for the purposes of a pilot. However, I'm hesitant about that because once this is enshrined in code it could contribute to wasteful use of resources on our servers (i.e., contribute to general slowness) for a long time before anyone ever revisits it.

Would a broader cache strategy work here? One idea is to save a flag in redis (or somewhere like that, maybe even postgres if redis is too unreliable) that indicates the last time locations have changed for the domain (the flag would be updated when any location is edited). If that key is missing then perform the full query and set the flag. If present, compare to `last_sync.date`... The advantage of that of approach is that it would be very inexpensive to check. The disadvantage is that it may result in some queries being performed that could be avoided by a more expensive caching strategy. This approach may not be useful if locations are edited frequently.

Reply:
Instead of introducing a new model or workflow would a UNION query work? https://docs.djangoproject.com/en/2.1/ref/models/querysets/#union something similar to (not sure if this exactly works, but hopefully conveys the idea):

```python
Location.objects.filter(domain=domain, date__gte=time).values('name').union(
    LocationType.filter(domain=domain, date__gte=time).values('name'), 
    RelatedLocation.filter(domain=domain, date__gte=time).values('name')).exists()
```

Reply:
I don't think you can mix different tables (model types) in a union query like that unless they have the same column names and types. Theoretically it could work if you only selected a single column (`id`, for example).

Reply:
That's what the `values('name')` was intended to represent.

anyways I followed up here: https://github.com/dimagi/commcare-hq/pull/23279